src/Agent.Worker/Variables.cs

```diff
@@ -589,6 +615,52 @@ public void CopyInto(Dictionary<string, VariableValue> target)
             }
         }
 
+        private Boolean IsReadOnly(Variable variable)
+        {
+            if (variable.ReadOnly)
+            {
+                return true;
+            }
+
+            // If variable is not marked as readOnly, check whether its in the set of reserved system variables (which are always readOnly)
+            if (_wellKnownSystemVariables == null)
+            {
+                _wellKnownSystemVariables = new List<String>();
+
+                // These are all classes which have system variables defined in them. See https://github.com/microsoft/azure-pipelines-agent/blob/master/src/Microsoft.VisualStudio.Services.Agent/Constants.cs
+                List<System.Type> wellKnownSystemVariableClasses = new List<System.Type>()
+                {
+                    typeof(Constants.Variables.Agent),
+                    typeof(Constants.Variables.Build),
+                    typeof(Constants.Variables.Features),
+                    typeof(Constants.Variables.Pipeline),
+                    typeof(Constants.Variables.Release),
+                    typeof(Constants.Variables.System),
+                    typeof(Constants.Variables.Task),
+                    typeof(WellKnownDistributedTaskVariables)
+                };
+
+                // Iterate through members of each class and add any system variables (aka prefixed with our readOnlyPrefixes)
+                foreach (System.Type systemVariableClass in wellKnownSystemVariableClasses)
+                {
+                    var wellKnownDistributedTaskFields = systemVariableClass.GetFields();
+                    foreach(var field in wellKnownDistributedTaskFields)
+                    {
+                        String value = field.GetValue(systemVariableClass).ToString().ToLower();
```

Comment:
if a field has a null value, does the ToString throw an exception?

Reply:
Yeah - in theory it should never happen, but I'll add a check.